### ISSUE TITLE
_check_conditional is broken on "string" type conditionals

### ISSUE THREAD

**Issue by DavidVentura** [ansible/ansible#41475]
**Please read at the end of the issue for an explanation and proposed solutions**

##### SUMMARY

The code that evaluates conditionals will simply return the value passed in if its type is string

##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME

ansible/playbook/conditional.py

##### ANSIBLE VERSION

```
ansible 2.5.4
  config file = None
  configured module search path = [u'/home/airflow/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/airflow/.venv/lib/python2.7/site-packages/ansible
  executable location = /home/airflow/.venv/bin/ansible
  python version = 2.7.5 (default, Jun 17 2014, 18:11:42) [GCC 4.8.2 20140120 (Red Hat 4.8.2-16)]

```

##### CONFIGURATION

nothing

##### OS / ENVIRONMENT

N/A
##### STEPS TO REPRODUCE

```yaml
- hosts: localhost
  gather_facts: no
  tasks:
  - name: do not fail me
    fail:
    when: false and true
```

##### EXPECTED RESULTS

playbook should not fail
##### ACTUAL RESULTS

playbook failed

```
N/A
```

The problem:
[Line 142 of conditional.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/conditional.py#L142)
```
            if not isinstance(conditional, text_type) or conditional == "":
                return conditional
```
where `text_type` here is `unicode`
If, for some reason you pass in a conditional with type string it will always return the string and if that has ANY content it will be evaluated to true

Even if this is not a condition that can be met using `ansible-playbook` (as it parses yaml as unicode), the code path where this has type string makes no sense so at least it should spit out a warning

Example:

```
u"true and false" = false
"true and false" = true
```

Proposed solution:

Give a warning or even fail execution when conditional is of type string; do not consider `string = true`

**Comment by ansibot** [comment]
Files identified in the description:
* [lib/ansible/playbook/conditional.py](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/conditional.py)

If these files are inaccurate, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by willthames** [comment]
This playbook doesn't fail for me on 2.5.0 or 2.5.4 or devel

```
$ ansible --version
ansible 2.5.4
  config file = /Users/will/tmp/ansible/ansible.cfg
  configured module search path = [u'/Users/will/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Mar  9 2018, 23:57:12) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```

Using your reproduction step playbook as 41475.yml, the `fail` task doesn't fail because the `when` clause evaluates to `False` as expected.

```
$ ansible-playbook 41475.yml 
 [WARNING]: Unable to parse /etc/ansible/hosts as an inventory source

 [WARNING]: No inventory was parsed, only implicit localhost is available

 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'


PLAY [localhost] ******************************************************************************************************************************************************************************

TASK [do not fail me] *************************************************************************************************************************************************************************
skipping: [localhost]

PLAY RECAP ************************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=0 
```

What am I doing differently?

**Comment by DavidVentura** [comment]
Nothing, as stated on the issue
`Even if this is not a condition that can (currently) be met using ansible-playbook, the code path where this has type string makes no sense at all, so at least it should spit out a warning`

**Comment by sivel** [comment]
I believe this is a duplicate of https://github.com/ansible/ansible/issues/34591 and https://github.com/ansible/ansible/issues/41236

Also, I think you are misinterpreting the line you indicate.  It says if it is *NOT* a string, go ahead and return it.

See https://github.com/ansible/ansible/issues/34591#issuecomment-356081973 for a description of what is happening.

**Comment by DavidVentura** [comment]
I'll add more details tomorrow but I debugged this with pdb today.

If "conditionals" is a (normal) string it'd do the weird behavior. If it is a unicode string then it has correct behavior.

**Comment by sivel** [comment]
Ah, so you are getting a "bytes" instance there somehow.

**Comment by DavidVentura** [comment]
Sorry that i confused you, it's a plain old string. Not bytes, not unicode. Yes, i am putting it there via an unsupported way, but still the behavior is extremely weird in that case

**Comment by sivel** [comment]
I should mention that in python2 that `string_types` == `(basestring,)` which matches native `str` and `unicode`.

On python3 `string_types` == `(str,)` which is the python3 equivalent of `unicode`, and we should never be dealing with byte strings in this part of the code in python3.

**Comment by DavidVentura** [comment]
Sorry for being extremely unclear, I'll show now the behavior on both cases; I argue that the behavior should either be the same or that at the very least you should get a big scary warning.

UNICODE:

[Line 145](https://github.com/ansible/ansible/blob/devel/lib/ansible/playbook/conditional.py#L145)
```
> ansible/playbook/conditional.py(145)_check_conditional()
-> if not isinstance(conditional, text_type) or conditional == "":
(Pdb) conditional
u'false and true
(Pdb) text_type
<type 'unicode'>
(Pdb) isinstance(conditional, text_type)
True
```

The condition on the if is not met, execution continues to evaluate the template.


STRING
```
> ansible/playbook/conditional.py(145)_check_conditional()
-> if not isinstance(conditional, text_type) or conditional == "":
(Pdb) conditional
'false and true'
(Pdb) text_type
<type 'unicode'>
(Pdb) isinstance(conditional, text_type)
False
(Pdb) n
> ansible/playbook/conditional.py(146)_check_conditional()
-> return conditional

```

Execution is "over". This will return the conditional and as long as it is not "" it will always be true; snippet of hitting 'n' after the last line

```
> ansible/playbook/conditional.py(146)_check_conditional()
-> return conditional
(Pdb) n
--Return--
> ansible/playbook/conditional.py(146)_check_conditional()->'false and true'
-> return conditional
(Pdb) n
> ansible/playbook/conditional.py(99)evaluate_conditional()
-> for conditional in self.when:
(Pdb) n
> ansible/playbook/conditional.py(107)evaluate_conditional()
-> return True
```

**Comment by abadger** [comment]
We probably want to figure out why we're getting a byte string there and convert it at the border instead of changing the conditional.

**Comment by abadger** [comment]
@DavidVentura Can you tell us how to reproduce this (ie, what playbook produces this error?)  That will greatly aid us in tracking down where the byte string is entering Ansible.

**Comment by DavidVentura** [comment]
As stated before, the playbook executor prevents this from happening, this is happening with a custom executor.

The issue is that if this variable is str the behavior is wrong. Just adding a warning or making the code follow the same path (with str) as the unicode path it would be ok

**Comment by bcoca** [comment]
@DavidVentura then the custom executor has a bug, we enforce the boundaries on input and transformation so this should always get the expected string format.

**Comment by DavidVentura** [comment]
Which I believe is OK, but shouldn't the cinditional evaluator emit a warning when it doesn't get the expected input type?

**Comment by bcoca** [comment]
I would raise an error, but it would never fire off unless you patch ansible itself.

**Comment by DavidVentura** [comment]
thing is, i've been developing a custom executor. Spent 2 days figuring out why some conditionals were always true. turns out my yaml parser was creating non-unicode strings, which TaskQueueManager gladly accepts, but then behaves weirdly.
I would like (as a dev) if at some point between TQM instantiation and conditional execution I could get SOME info about my input being "invalid"

**Comment by bcoca** [comment]
this is why we control the input boundaries, otherwise we would have to have many more checks at every stage of the code cause we could never trust what we are getting.

**Comment by sivel** [comment]
> I would like (as a dev) if at some point between TQM instantiation and conditional execution I could get SOME info about my input being "invalid"

Just as an FYI, in case no one has mentioned it yet, but utilizing the Playbook or task execution Python APIs within Ansible is unsupported.  Those APIs are private, and not intended for public consumption.

**Comment by sivel** [comment]
Internal Ansible API
====================

Hi!

Thanks very much for your submission to Ansible.  It sincerely means a lot to us
to have an active community that is willing to submit feedback so that we may
continue to make Ansible better.

The internal Python API is considered an unsupported aspect of Ansible and this
is not considered a bug unless there is an issue with an Ansible binary
(`ansible`, `ansible-playbook`, `ansible-doc`, etc) or an issue with an external
API such as are provided for the development of plugins (modules, dynamic
inventories, callbacks, strategies, etc).

Thank you once again for this and your interest in Ansible!

**Comment by abadger** [comment]
I took a look at the site in question.  It doesn't look like we could check
there as the logic is, "if not a text string, pass this on for something
else to try to handle.". There may be somewhere further up the call stack
where we can check and convert if a byte string is given.

However, checking has a cost.  Getting rid of extra code that checks and
converts at every step of the process is good for ansible.  Perhaps what we
should do is document that all strings should be text strings unless the
functions' docstring states otherwise.  The only problem is that ansible's
internal API isn't currently well documented (it's a low priority compared
to documenting the public interfaces as well, so not likely to happen for a
while) so people might not even read the documentation and see that note.

Did you read any documentation before starting to write your custom
executor?  If so, that's probably a good place for me to either wire this
down our link to it if I document it elsewhere (for instance, in our coding
guidelines)
